### PR TITLE
Feautre: Better handling for stop signal

### PR DESCRIPTION
**Pull request by noam-shakuri**

### CODE REVIEW

File: docs/worker/config/logging.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ There are the following configuration options:
     - `on_retry` [True] - Whenever a task asks for retry.
     - `on_max_retries` [True] - Each time a task asks for a retry beyond the maximum number of tries.
     - `on_requeue` [True] - Every time a task requeues.
+    - `on_stop` [True] - Every time a task stopped on the middle.
```
Comment (wavenator) [inline docs/worker/config/logging.md:46]: Every time a task **is** stopped **in** the middle.

I prefer a different phrase such as:
Every time a task calls `stop` or is interrupted by a `SIGTERM`.

File: docs/worker/handlers/on_stop.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# on_stop
+
+The `on_stop` handler is invoked when a task called the `stop` method or a when a SIGTERM signal received.
```
Comment (wavenator) [inline docs/worker/handlers/on_stop.md:3]: The `on_stop` handler is invoked when a task calls the `stop` method or receives a `SIGTERM` signal.

File: docs/worker/methods/stop.md [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 It is possible to stop the worker from running by using the `stop` method. The supervisor receives the request, and will never spawn a new worker instead. A worker uses this method when there is no solution to a problem. Using this method in conjunction with `on_starvation` is a good example of how it should be used. If a worker is starving and there are not enough tasks available to consume, it can be intentionally stopped to reduce the load on the queue. By calling `stop`, all the workers under a specific supervisor will stop as well.
 
+Also, when a SIGTERM signal is received, the `stop` function is being called.
```
Comment (wavenator) [inline docs/worker/methods/stop.md:5]: When the worker is interrupted by a `SIGTERM`, the signal handler invokes the `stop` function to interrupt the worker's workflow.

File: docs/worker/methods/stop.md [hunk 4]
```diff
@@ -53,4 +55,13 @@ def stop(
     ):
         if isinstance(exception, pymongo.errors.ServerSelectionTimeoutError):
             self.stop()
+
+    def on_stop(
+        self,
+        task,
+    ):
+        super().on_stop(
```
Comment (wavenator) [inline docs/worker/methods/stop.md:63]: There is no need to call the parent's `on_stop` as it is just an empty stub.

File: sergeant/executor/threaded.py [hunk 5]
```diff
@@ -133,6 +133,17 @@ def execute_task(
             self.worker_object.handle_requeue(
                 task=task,
             )
+        except worker.WorkerStop as exception:
```
Comment (wavenator) [inline sergeant/executor/threaded.py:136]: You should streamline the stop signal handling under `except worker.WorkerInterrupt as exception:` on line 147. WorkerStop exception derives from WorkerInterrupt exception class. The second option is to unpack the WorkerInterrupt handling into its parts and leave this function as is but marking the reason for the interruption under `self.interrupt_exception` variable

File: sergeant/worker.py [hunk 6]
```diff
@@ -615,6 +621,34 @@ def handle_starvation(
                 },
             )
 
+    def handle_stop(
+        self,
+        task: objects.Task,
+    ) -> None:
+        try:
+            if self.config.logging.events.on_stop:
+                self.logger.error(
```
Comment (wavenator) [inline sergeant/worker.py:630]: A task asking for a stop is not an error. You should change the logging level to either warning or info

File: docs/worker/methods/stop.md (hunk 4 above)
Comment (noam-shakuri) [reply docs/worker/methods/stop.md:63]: not exactly.. It raise an `WorkerStop` exception which stop the worker

Comment (wavenator) [reply docs/worker/methods/stop.md:63]: WDYM? the base on_stop is just an empty stub. What am I missing?

**Review by wavenator — APPROVED** [review]

**Pull request merged**